File: p2p-ws-sdk-dart/tool/verify_vectors.dart
import "dart:convert";
import "dart:io";
import "dart:typed_data";

import "package:convert/convert.dart";

import "../lib/src/frame.dart";
import "../lib/src/xor.dart";

Future<void> main() async {
  final here = File.fromUri(Platform.script).parent;
  final sdkRoot = here.parent;
  final vectorsDir = Directory("${sdkRoot.parent.path}/p2p-ws-protocol/test-vectors");

  await _verifyFrameVector(File("${vectorsDir.path}/frame_vector_001.json"));
  await _verifyXorVector(File("${vectorsDir.path}/xor_vector_001.json"));
  stdout.writeln("OK");
}

Future<void> _verifyFrameVector(File f) async {
  final j = jsonDecode(await f.readAsString()) as Map<String, dynamic>;
  final header = j["header"] as Map<String, dynamic>;
  final lengthU32 = (header["length_u32"] as num).toInt();
  final magicU16 = (header["magic_u16"] as num).toInt();
  final versionU8 = (header["version_u8"] as num).toInt();
  final flagsU8 = (header["flags_u8"] as num).toInt();
  final cipherPayload = Uint8List.fromList(hex.decode(j["cipher_payload_hex"] as String));
  final expectedWs = hex.decode(j["ws_binary_payload_hex"] as String);

  final got = encodeFrame(WireHeader(lengthU32, magicU16, versionU8, flagsU8), cipherPayload);
  final gotHex = hex.encode(got);
  final expHex = hex.encode(expectedWs);
  if (gotHex != expHex) {
    throw StateError("frame vector mismatch: got=$gotHex expected=$expHex");
  }

  final decoded = decodeFrame(got);
  if (decoded.header.length != lengthU32) throw StateError("frame decode length mismatch");
  if (decoded.header.magic != magicU16) throw StateError("frame decode magic mismatch");
  if (decoded.header.version != versionU8) throw StateError("frame decode version mismatch");
  if (decoded.header.flags != flagsU8) throw StateError("frame decode flags mismatch");
  if (hex.encode(decoded.cipherPayload) != hex.encode(cipherPayload)) throw StateError("frame decode payload mismatch");
}

Future<void> _verifyXorVector(File f) async {
  final j = jsonDecode(await f.readAsString()) as Map<String, dynamic>;
  final plain = Uint8List.fromList(hex.decode(j["plain_hex"] as String));
  final keyfile = Uint8List.fromList(hex.decode(j["keyfile_bytes_hex"] as String));
  final offset = (j["offset"] as num).toInt();
  final keySlice = Uint8List.sublistView(keyfile, offset, offset + plain.length);
  final expectedCipher = hex.decode(j["cipher_hex"] as String);
  final gotCipher = xorNoWrap(plain, keySlice, 0);
  if (hex.encode(gotCipher) != hex.encode(expectedCipher)) {
    throw StateError("xor vector mismatch");
  }
  final back = xorNoWrap(gotCipher, keySlice, 0);
  if (hex.encode(back) != hex.encode(plain)) {
    throw StateError("xor invert mismatch");
  }
}

File: p2p-ws-protocol/scripts/verify_vectors.py
import json
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
VEC_DIR = ROOT / "test-vectors"


def hx(s: str) -> bytes:
    s = s.strip()
    return bytes.fromhex(s)


def xor_bytes(key: bytes, offset: int, plain: bytes) -> bytes:
    if offset < 0 or offset >= len(key):
        raise ValueError("offset out of range")
    if offset + len(plain) > len(key):
        raise ValueError("offset+plainLen exceeds keyLen (no wrap)")
    out = bytearray(len(plain))
    for i, b in enumerate(plain):
        out[i] = b ^ key[offset + i]
    return bytes(out)

def decode_header(frame: bytes) -> tuple[int, int, int, int]:
    if len(frame) < 8:
        raise ValueError("invalid frame")
    length = int.from_bytes(frame[0:4], "big", signed=False)
    magic = int.from_bytes(frame[4:6], "big", signed=False)
    version = frame[6]
    flags = frame[7]
    return length, magic, version, flags


def main() -> int:
    vectors = sorted(VEC_DIR.glob("xor_vector_*.json"))
    if not vectors:
        raise SystemExit("no xor_vector_*.json found")
    ok = 0
    for p in vectors:
        v = json.loads(p.read_text(encoding="utf-8"))
        key = hx(v["keyfile_bytes_hex"])
        offset = int(v["offset"])
        plain = hx(v["plain_hex"])
        expected = v["cipher_hex"].lower()
        got = xor_bytes(key, offset, plain).hex()
        if got != expected:
            raise SystemExit(f"{p.name}: mismatch: got={got} expected={expected}")
        ok += 1

    frame_vectors = sorted(VEC_DIR.glob("frame_vector_*.json"))
    for p in frame_vectors:
        v = json.loads(p.read_text(encoding="utf-8"))
        ws = hx(v["ws_binary_payload_hex"])
        length, magic, version, flags = decode_header(ws)
        h = v["header"]
        if length != int(h["length_u32"]) or magic != int(h["magic_u16"]) or version != int(h["version_u8"]) or flags != int(h["flags_u8"]):
            raise SystemExit(f"{p.name}: header mismatch")
        cipher = ws[8:].hex()
        expected_cipher = str(v["cipher_payload_hex"]).lower()
        if cipher != expected_cipher:
            raise SystemExit(f"{p.name}: payload mismatch: got={cipher} expected={expected_cipher}")
        ok += 1

    print(f"ok={ok}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
